File: HisabCore/Sources/HisabCore/GenericBank/ColumnInference.swift
import Foundation

/// Conservative column inference: enumerates candidate mappings for an unknown
/// bank table and accepts a result only when exactly one distinct interpretation
/// closes the balance chain. Zero candidates or conflicting winners mean nil —
/// the engine never guesses.
public enum ColumnInference {
    public struct Result: Sendable {
        public let mapping: ColumnMapping
        public let transactions: [ParsedTransaction]
    }

    static let candidateDateFormats = ["dd/MM/yyyy", "dd/MM/yy", "dd-MM-yyyy",
                                       "dd MMM yyyy", "dd-MMM-yyyy", "dd-MMM-yy",
                                       "yyyy-MM-dd"]

    public static func infer(table: NormalizedTable) -> Result? {
        let headerIndex = detectHeader(rows: table.rows)
        let start = headerIndex.map { $0 + 1 } ?? 0
        var body: [[String]] = []
        for row in table.rows.dropFirst(start) {
            let nonEmpty = row.filter { !$0.isEmpty }.count
            if nonEmpty >= 3 { body.append(row) }
        }
        guard body.count >= 2 else { return nil }

        let width = body.map(\.count).max() ?? 0
        guard width >= 3 else { return nil }

        func cells(_ column: Int) -> [String] {
            body.map { column < $0.count ? $0[column] : "" }
        }

        // Classify columns.
        var dateColumns: [(column: Int, format: String)] = []
        var numericFull: [Int] = []      // every row numeric — balance/amount candidates
        var numericSparse: [Int] = []    // numeric or empty, ≥1 non-empty — debit/credit legs
        var drcrColumns: [Int] = []
        var textColumns: [(column: Int, averageLength: Int)] = []

        for column in 0..<width {
            let values = cells(column)
            if let format = dateFormat(parsingAll: values) {
                dateColumns.append((column, format))
                continue
            }
            let nonEmpty = values.filter { !$0.isEmpty }
            let numericCount = nonEmpty.filter { Money.signedPaise(fromDecimalString: $0) != nil }.count
            if !nonEmpty.isEmpty && numericCount == nonEmpty.count {
                if nonEmpty.count == values.count { numericFull.append(column) }
                numericSparse.append(column)
                continue
            }
            let drcrCount = nonEmpty.filter { ["dr", "cr"].contains($0.lowercased()) }.count
            if !nonEmpty.isEmpty && drcrCount == nonEmpty.count {
                drcrColumns.append(column)
                continue
            }
            if !nonEmpty.isEmpty {
                let total = nonEmpty.reduce(0) { $0 + $1.count }
                textColumns.append((column, total / nonEmpty.count))
            }
        }

        guard let (dateColumn, format) = dateColumns.first, dateColumns.count == 1 else {
            // Multiple date columns (Value Date + Txn Date) are fine — prefer the
            // leftmost; zero means no bank table.
            if let first = dateColumns.first {
                return inferWithDate(dateColumn: first.column, format: first.format,
                                     numericFull: numericFull, numericSparse: numericSparse,
                                     drcrColumns: drcrColumns, textColumns: textColumns,
                                     body: body)
            }
            return nil
        }
        return inferWithDate(dateColumn: dateColumn, format: format,
                             numericFull: numericFull, numericSparse: numericSparse,
                             drcrColumns: drcrColumns, textColumns: textColumns,
                             body: body)
    }

    private static func inferWithDate(dateColumn: Int, format: String,
                                      numericFull: [Int], numericSparse: [Int],
                                      drcrColumns: [Int],
                                      textColumns: [(column: Int, averageLength: Int)],
                                      body: [[String]]) -> Result? {
        // Narration: the longest-averaging text column.
        let narration = textColumns.max(by: { $0.averageLength < $1.averageLength })?.column
        guard let narrationColumn = narration else { return nil }
        // Reference: any other text column (short alphanumeric labels).
        let referenceColumn = textColumns
            .filter { $0.column != narrationColumn }
            .min(by: { $0.averageLength < $1.averageLength })?.column

        var mappings: [ColumnMapping] = []
        for balance in numericFull {
            let others = numericSparse.filter { $0 != balance }
            // (a) separate debit/credit legs
            for debit in others {
                for credit in others where credit != debit {
                    mappings.append(ColumnMapping(date: dateColumn, narration: narrationColumn,
                                                  reference: referenceColumn,
                                                  debit: debit, credit: credit,
                                                  balance: balance, dateFormat: format))
                }
            }
            // (b) single signed amount / (c) amount + DR/CR
            for amount in others where numericFull.contains(amount) {
                mappings.append(ColumnMapping(date: dateColumn, narration: narrationColumn,
                                              reference: referenceColumn,
                                              amount: amount,
                                              balance: balance, dateFormat: format))
                for drcr in drcrColumns {
                    mappings.append(ColumnMapping(date: dateColumn, narration: narrationColumn,
                                                  reference: referenceColumn,
                                                  amount: amount, drcr: drcr,
                                                  balance: balance, dateFormat: format))
                }
            }
        }

        var winners: [(ColumnMapping, [ParsedTransaction])] = []
        for mapping in mappings {
            if case .validated(let txns) = ChainInterpreter.interpret(rows: body, mapping: mapping) {
                winners.append((mapping, txns))
            }
        }
        guard let first = winners.first else { return nil }
        for (_, txns) in winners.dropFirst() where txns != first.1 {
            return nil  // two different valid readings — refuse to choose
        }
        return Result(mapping: first.0, transactions: first.1)
    }

    /// The first row that names columns rather than containing data: ≥2 cells
    /// matching header vocabulary and no date-parseable cell.
    static func detectHeader(rows: [[String]]) -> Int? {
        let keywords = try! NSRegularExpression(
            pattern: "date|narration|particular|description|remarks|debit|credit|withdraw|deposit|amount|balance|ref|cheque",
            options: [.caseInsensitive])
        for (index, row) in rows.enumerated() {
            var keywordHits = 0
            var hasDate = false
            for cell in row where !cell.isEmpty {
                let range = NSRange(cell.startIndex..., in: cell)
                if keywords.firstMatch(in: cell, range: range) != nil { keywordHits += 1 }
                if dateFormat(parsingAll: [cell]) != nil { hasDate = true }
            }
            if keywordHits >= 2 && !hasDate { return index }
        }
        return nil
    }

    /// First candidate format under which every non-empty value parses (and at
    /// least one value is non-empty). Nil when the column isn't dates.
    private static func dateFormat(parsingAll values: [String]) -> String? {
        let formatter = DateFormatter()
        formatter.calendar = YearMonth.istCalendar
        formatter.timeZone = YearMonth.istCalendar.timeZone
        formatter.locale = Locale(identifier: "en_US_POSIX")
        let nonEmpty = values.filter { !$0.isEmpty }
        guard !nonEmpty.isEmpty else { return nil }
        for candidate in candidateDateFormats {
            formatter.dateFormat = candidate
            var allParse = true
            for value in nonEmpty where formatter.date(from: value) == nil {
                allParse = false
                break
            }
            if allParse { return candidate }
        }
        return nil
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/generic_bank/column_inference.dart
/// Conservative column inference. Port of ColumnInference.swift — accepts a
/// result only when exactly one distinct interpretation closes the balance
/// chain; zero candidates or conflicting winners mean null. Never guesses.
library;

import '../money.dart';
import '../domain.dart';
import 'chain_interpreter.dart';
import 'normalized_table.dart';
import 'statement_date.dart';

class InferenceResult {
  final ColumnMapping mapping;
  final List<ParsedTransaction> transactions;
  const InferenceResult(this.mapping, this.transactions);
}

class ColumnInference {
  static const candidateDateFormats = [
    'dd/MM/yyyy', 'dd/MM/yy', 'dd-MM-yyyy',
    'dd MMM yyyy', 'dd-MMM-yyyy', 'dd-MMM-yy', 'yyyy-MM-dd',
  ];

  static InferenceResult? infer(NormalizedTable table) {
    final headerIndex = detectHeader(table.rows);
    final start = headerIndex == null ? 0 : headerIndex + 1;
    final body = <List<String>>[];
    for (final row in table.rows.skip(start)) {
      final nonEmpty = row.where((c) => c.isNotEmpty).length;
      if (nonEmpty >= 3) body.add(row);
    }
    if (body.length < 2) return null;

    var width = 0;
    for (final row in body) {
      if (row.length > width) width = row.length;
    }
    if (width < 3) return null;

    List<String> cells(int column) =>
        body.map((r) => column < r.length ? r[column] : '').toList();

    final dateColumns = <(int, String)>[];
    final numericFull = <int>[];
    final numericSparse = <int>[];
    final drcrColumns = <int>[];
    final textColumns = <(int, int)>[]; // (column, averageLength)

    for (var column = 0; column < width; column++) {
      final values = cells(column);
      final format = _dateFormatParsingAll(values);
      if (format != null) {
        dateColumns.add((column, format));
        continue;
      }
      final nonEmpty = values.where((v) => v.isNotEmpty).toList();
      final numericCount =
          nonEmpty.where((v) => Money.signedPaise(v) != null).length;
      if (nonEmpty.isNotEmpty && numericCount == nonEmpty.length) {
        if (nonEmpty.length == values.length) numericFull.add(column);
        numericSparse.add(column);
        continue;
      }
      final drcrCount = nonEmpty
          .where((v) => v.toLowerCase() == 'dr' || v.toLowerCase() == 'cr')
          .length;
      if (nonEmpty.isNotEmpty && drcrCount == nonEmpty.length) {
        drcrColumns.add(column);
        continue;
      }
      if (nonEmpty.isNotEmpty) {
        final total = nonEmpty.fold<int>(0, (acc, v) => acc + v.length);
        textColumns.add((column, total ~/ nonEmpty.length));
      }
    }

    if (dateColumns.isEmpty) return null;
    // Multiple date columns (Value Date + Txn Date): prefer the leftmost.
    final (dateColumn, format) = dateColumns.first;
    return _inferWithDate(dateColumn, format, numericFull, numericSparse,
        drcrColumns, textColumns, body);
  }

  static InferenceResult? _inferWithDate(
      int dateColumn,
      String format,
      List<int> numericFull,
      List<int> numericSparse,
      List<int> drcrColumns,
      List<(int, int)> textColumns,
      List<List<String>> body) {
    if (textColumns.isEmpty) return null;
    var narrationColumn = textColumns.first.$1;
    var bestLength = textColumns.first.$2;
    for (final (column, avg) in textColumns) {
      if (avg > bestLength) {
        narrationColumn = column;
        bestLength = avg;
      }
    }
    int? referenceColumn;
    var shortest = 1 << 30;
    for (final (column, avg) in textColumns) {
      if (column == narrationColumn) continue;
      if (avg < shortest) {
        referenceColumn = column;
        shortest = avg;
      }
    }

    final mappings = <ColumnMapping>[];
    for (final balance in numericFull) {
      final others = numericSparse.where((c) => c != balance).toList();
      for (final debit in others) {
        for (final credit in others) {
          if (credit == debit) continue;
          mappings.add(ColumnMapping(
              date: dateColumn,
              narration: narrationColumn,
              reference: referenceColumn,
              debit: debit,
              credit: credit,
              balance: balance,
              dateFormat: format));
        }
      }
      for (final amount in others) {
        if (!numericFull.contains(amount)) continue;
        mappings.add(ColumnMapping(
            date: dateColumn,
            narration: narrationColumn,
            reference: referenceColumn,
            amount: amount,
            balance: balance,
            dateFormat: format));
        for (final drcr in drcrColumns) {
          mappings.add(ColumnMapping(
              date: dateColumn,
              narration: narrationColumn,
              reference: referenceColumn,
              amount: amount,
              drcr: drcr,
              balance: balance,
              dateFormat: format));
        }
      }
    }

    ColumnMapping? winnerMapping;
    List<ParsedTransaction>? winnerTxns;
    for (final mapping in mappings) {
      final outcome = ChainInterpreter.interpret(rows: body, mapping: mapping);
      if (outcome is! ChainValidated) continue;
      if (winnerTxns == null) {
        winnerMapping = mapping;
        winnerTxns = outcome.transactions;
      } else if (!_sameTxns(winnerTxns, outcome.transactions)) {
        return null; // two different valid readings — refuse to choose
      }
    }
    if (winnerMapping == null || winnerTxns == null) return null;
    return InferenceResult(winnerMapping, winnerTxns);
  }

  static bool _sameTxns(List<ParsedTransaction> a, List<ParsedTransaction> b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  /// The first row that names columns rather than containing data: ≥2 cells
  /// matching header vocabulary and no date-parseable cell.
  static int? detectHeader(List<List<String>> rows) {
    final keywords = RegExp(
        'date|narration|particular|description|remarks|debit|credit|withdraw|deposit|amount|balance|ref|cheque',
        caseSensitive: false);
    for (var index = 0; index < rows.length; index++) {
      var keywordHits = 0;
      var hasDate = false;
      for (final cell in rows[index]) {
        if (cell.isEmpty) continue;
        if (keywords.hasMatch(cell)) keywordHits++;
        if (_dateFormatParsingAll([cell]) != null) hasDate = true;
      }
      if (keywordHits >= 2 && !hasDate) return index;
    }
    return null;
  }

  static String? _dateFormatParsingAll(List<String> values) {
    final nonEmpty = values.where((v) => v.isNotEmpty).toList();
    if (nonEmpty.isEmpty) return null;
    for (final candidate in candidateDateFormats) {
      var allParse = true;
      for (final value in nonEmpty) {
        if (StatementDate.tryParse(value, candidate) == null) {
          allParse = false;
          break;
        }
      }
      if (allParse) return candidate;
    }
    return null;
  }
}
